app/models/workers_scaling/algorithm.rb

```diff
@@ -0,0 +1,109 @@
+require_relative 'experiment_resources_interface'
+require_relative 'experiment_statistics'
+module WorkersScaling
+  ##
+  # Class describing interface of Workers scaling algorithm. Creating new algorithm
+  # requires providing class with proper methods implemented, which also should
+  # inherit Algorithm class.
+  class Algorithm < Scalarm::Database::MongoActiveRecord
+    use_collection 'workers_scaling_algorithms'
+    attr_accessor :experiment
+    attr_accessor :resources_interface
+    attr_accessor :experiment_statistics
+
+    NOT_IMPLEMENTED = 'This is an abstract method, which must be implemented by all subclasses'
+    ALGORITHM_INTERVAL = 30.seconds
+    ERRORS_MAX = 3
+
+    ##
+    # Returns name of Algorithm implementation class
+    def self.get_class_name
+      self.name.gsub('::', '__').underscore.to_sym
+    end
+
+    ##
+    # Arguments: attributes hash containing fields:
+    #  * experiment_id - id of Experiment to be subjected to Algorithm
+    #  * user_id - id of User starting Algorithm
+    #  * allowed_infrastructures - list of hashes with infrastructure and maximal Workers amount
+    #      (Detailed description at ExperimentResourcesInterface#initialize)
+    #  * planned_finish_time - desired time of end of Experiment (as Time instance)
+    #  * last_update_time - time of last change of user-defined fields (allowed_infrastructures, planned_finish_time)
+    #  * params (optional) - additional params, currently unused, may be used in subclasses
+    # All these fields are available in any Algorithm as if attr_accessor was created for each of them
+    def initialize(attributes)
+      super(attributes)
+    end
+
+    ##
+    # Must be executed before running #initial_deployment or #experiment_status_check
+    # Initializes fields that are not stored in database:
+    #  * @experiment
+    #  * @resources_interface
+    #  * @experiment_statistics
+    # Returns self to allow chaining
+    def initialize_runtime_fields
+      @experiment = Experiment.find_by_id(experiment_id)
```

Comment:
Użycie ``attr_accessor :experiment`` na początku powoduje, że mamy metodę ``experiment``, która sama zajmuje się cache'owaniem i wykonuje ``@experiment ||= Experiment.find_by_id(...)`` tylko raz dla instancji. Ta metoda też używa wewnętrznie ``@experiment``. Także to nie było potrzebne, chociaż nie powinno nic psuć.